From ceea21377fd15e2b96a6b3425df40ed246b4bd5d Mon Sep 17 00:00:00 2001
From: Evoke Zhang <evoke.zhang@amlogic.com>
Date: Tue, 9 Sep 2014 16:50:35 +0800
Subject: [PATCH 4838/5965] PD#97121: lcd driver: fix clock calculation mistake
 when set frame_rate as pclk

Squashed commit of the following:

commit 840d8de7d16b24925e2986641aceba6bce8e5628
Author: Evoke Zhang <evoke.zhang@amlogic.com>
Date:   Thu Sep 4 18:38:21 2014 +0800

    PD#97121: lcd driver: fix m6 compile error

    Change-Id: I7786bae33497daa54d588fdc9d119ec60d8f9783

commit 28c7cb4edb3cd5fd7bf497735358038777286393
Author: Evoke Zhang <evoke.zhang@amlogic.com>
Date:   Thu Sep 4 14:46:20 2014 +0800

    PD#97121: lcd driver: update driver version

    Change-Id: I8af61ca0937c6e23313ced1275a0f4acc7e3cad4

commit 39b634030816b17592c215075439021e42f838aa
Author: Evoke Zhang <evoke.zhang@amlogic.com>
Date:   Thu Sep 4 14:01:19 2014 +0800

    PD#97121: lcd driver: fix clock calculation mistake when set frame_rate as pclk

    Change-Id: I8cf6b2c9db44dfc62fe4cc9de1bb04654fac0a84

Change-Id: I405c85065e55d3c1e2610b6ca5187db968849c0f
---
 arch/arm/mach-meson6/lcd/lcd_config.c  | 15 +++++++--------
 arch/arm/mach-meson6/lcd/lcd_config.h  |  2 +-
 arch/arm/mach-meson8/lcd/lcd_config.c  | 15 +++++++--------
 arch/arm/mach-meson8/lcd/lcd_config.h  |  2 +-
 arch/arm/mach-meson8b/lcd/lcd_config.c | 15 +++++++--------
 arch/arm/mach-meson8b/lcd/lcd_config.h |  2 +-
 6 files changed, 24 insertions(+), 27 deletions(-)

diff --git a/arch/arm/mach-meson6/lcd/lcd_config.c b/arch/arm/mach-meson6/lcd/lcd_config.c
index 28890f394b11..b08daaa74ee3 100755
--- a/arch/arm/mach-meson6/lcd/lcd_config.c
+++ b/arch/arm/mach-meson6/lcd/lcd_config.c
@@ -1562,15 +1562,10 @@ static void generate_clk_parameter(Lcd_Config_t *pConf)
     unsigned min_error = MAX_ERROR;
     unsigned error = MAX_ERROR;
     unsigned clk_num = 0;
-    unsigned fin = FIN_FREQ;
-    unsigned fout = pConf->lcd_timing.lcd_clk;
+    unsigned fin, fout;
 
-    if (fout >= 200) {//clk
-        fout = fout / 1000;  //kHz
-    }
-    else {//frame_rate
-        fout = (fout * pConf->lcd_basic.h_period * pConf->lcd_basic.v_period) / 1000;	//kHz
-    }
+    fin = FIN_FREQ; //kHz
+    fout = pConf->lcd_timing.lcd_clk / 1000; //kHz
 
     switch (pConf->lcd_basic.lcd_type) {
         case LCD_DIGITAL_LVDS:
@@ -1814,6 +1809,10 @@ static void lcd_control_config_pre(Lcd_Config_t *pConf)
 {
     unsigned ss_level;
 
+    if (pConf->lcd_timing.lcd_clk < 200) {//prepare refer clock for frame_rate setting
+        pConf->lcd_timing.lcd_clk = (pConf->lcd_timing.lcd_clk * pConf->lcd_basic.h_period * pConf->lcd_basic.v_period);
+    }
+
     ss_level = ((pConf->lcd_timing.clk_ctrl >> CLK_CTRL_SS) & 0xf);
     ss_level = ((ss_level >= SS_LEVEL_MAX) ? (SS_LEVEL_MAX-1) : ss_level);
 
diff --git a/arch/arm/mach-meson6/lcd/lcd_config.h b/arch/arm/mach-meson6/lcd/lcd_config.h
index 527ec6422bd4..95009cb0addb 100755
--- a/arch/arm/mach-meson6/lcd/lcd_config.h
+++ b/arch/arm/mach-meson6/lcd/lcd_config.h
@@ -8,7 +8,7 @@
 //lcd driver version
 //**********************************
 #define LCD_DRV_TYPE      "c6"
-#define LCD_DRV_DATE      "20140826"
+#define LCD_DRV_DATE      "20140903"
 //**********************************
 
 /* for GAMMA_CNTL_PORT */
diff --git a/arch/arm/mach-meson8/lcd/lcd_config.c b/arch/arm/mach-meson8/lcd/lcd_config.c
index dbc7e965f117..db68ee0c83c3 100755
--- a/arch/arm/mach-meson8/lcd/lcd_config.c
+++ b/arch/arm/mach-meson8/lcd/lcd_config.c
@@ -1317,15 +1317,10 @@ static void generate_clk_parameter(Lcd_Config_t *pConf)
     unsigned edp_tx_phy_out;
     unsigned clk_num = 0;
     unsigned tmp;
-    unsigned fin = FIN_FREQ;
-    unsigned fout = pConf->lcd_timing.lcd_clk;
+    unsigned fin, fout;
 
-    if (fout >= 200) {//clk
-        fout = fout / 1000;  //kHz
-    }
-    else {//frame_rate
-        fout = (fout * pConf->lcd_basic.h_period * pConf->lcd_basic.v_period) / 1000;	//kHz
-    }
+    fin = FIN_FREQ; //kHz
+    fout = pConf->lcd_timing.lcd_clk / 1000; //kHz
 
     switch (pConf->lcd_basic.lcd_type) {
         case LCD_DIGITAL_MIPI:
@@ -1733,6 +1728,10 @@ static void lcd_control_config_pre(Lcd_Config_t *pConf)
 {
     unsigned ss_level;
 
+    if (pConf->lcd_timing.lcd_clk < 200) {//prepare refer clock for frame_rate setting
+        pConf->lcd_timing.lcd_clk = (pConf->lcd_timing.lcd_clk * pConf->lcd_basic.h_period * pConf->lcd_basic.v_period);
+    }
+
     ss_level = ((pConf->lcd_timing.clk_ctrl >> CLK_CTRL_SS) & 0xf);
     ss_level = ((ss_level >= SS_LEVEL_MAX) ? (SS_LEVEL_MAX-1) : ss_level);
 
diff --git a/arch/arm/mach-meson8/lcd/lcd_config.h b/arch/arm/mach-meson8/lcd/lcd_config.h
index 2c4e5d4209bf..e6b12ca948c4 100755
--- a/arch/arm/mach-meson8/lcd/lcd_config.h
+++ b/arch/arm/mach-meson8/lcd/lcd_config.h
@@ -7,7 +7,7 @@
 //lcd driver version
 //**********************************
 #define LCD_DRV_TYPE      "c8"
-#define LCD_DRV_DATE      "20140826"
+#define LCD_DRV_DATE      "20140903"
 //**********************************
 
 /* for GAMMA_CNTL_PORT */
diff --git a/arch/arm/mach-meson8b/lcd/lcd_config.c b/arch/arm/mach-meson8b/lcd/lcd_config.c
index 4975a48a7778..81b89f8646e3 100755
--- a/arch/arm/mach-meson8b/lcd/lcd_config.c
+++ b/arch/arm/mach-meson8b/lcd/lcd_config.c
@@ -999,15 +999,10 @@ static void generate_clk_parameter(Lcd_Config_t *pConf)
     unsigned int dsi_bit_rate_min=0, dsi_bit_rate_max=0;
     unsigned clk_num = 0;
     unsigned tmp;
-    unsigned fin = FIN_FREQ;
-    unsigned fout = pConf->lcd_timing.lcd_clk;
+    unsigned fin, fout;
 
-    if (fout >= 200) {//clk
-        fout = fout / 1000;  //kHz
-    }
-    else {//frame_rate
-        fout = (fout * pConf->lcd_basic.h_period * pConf->lcd_basic.v_period) / 1000;	//kHz
-    }
+    fin = FIN_FREQ; //kHz
+    fout = pConf->lcd_timing.lcd_clk / 1000; //kHz
 
     switch (pConf->lcd_basic.lcd_type) {
         case LCD_DIGITAL_MIPI:
@@ -1334,6 +1329,10 @@ static void lcd_control_config_pre(Lcd_Config_t *pConf)
 {
     unsigned ss_level;
 
+    if (pConf->lcd_timing.lcd_clk < 200) {//prepare refer clock for frame_rate setting
+        pConf->lcd_timing.lcd_clk = (pConf->lcd_timing.lcd_clk * pConf->lcd_basic.h_period * pConf->lcd_basic.v_period);
+    }
+
     ss_level = ((pConf->lcd_timing.clk_ctrl >> CLK_CTRL_SS) & 0xf);
     ss_level = ((ss_level >= SS_LEVEL_MAX) ? (SS_LEVEL_MAX-1) : ss_level);
 
diff --git a/arch/arm/mach-meson8b/lcd/lcd_config.h b/arch/arm/mach-meson8b/lcd/lcd_config.h
index 026326c21653..1696cf05dc36 100755
--- a/arch/arm/mach-meson8b/lcd/lcd_config.h
+++ b/arch/arm/mach-meson8b/lcd/lcd_config.h
@@ -7,7 +7,7 @@
 //lcd driver version
 //**********************************
 #define LCD_DRV_TYPE      "c8b"
-#define LCD_DRV_DATE      "20140826"
+#define LCD_DRV_DATE      "20140903"
 //**********************************
 
 /* for GAMMA_CNTL_PORT */
-- 
2.19.0

